rust/lancedb/src/arrow.rs

```diff
@@ -119,3 +128,214 @@ impl<T: arrow_array::RecordBatchReader + Send + 'static> IntoArrow for T {
         Ok(Box::new(self))
     }
 }
+
+#[cfg(feature = "polars")]
+/// An iterator of record batches formed from a Polars DataFrame. The iterator
+/// panics if the DataFrame's chunks are not aligned. Consider calling
+/// [`polars::prelude::DataFrame::should_rechunk`] to determine whether the DataFrame
+/// should be rechunked and calling [`polars::prelude::DataFrame::align_chunks`]
+/// if the DataFrame needs to be rechunked before using the iterator.
+pub struct PolarsDataFrameRecordBatchReader {
+    chunks: Vec<ArrowChunk>,
+    arrow_schema: Arc<arrow_schema::Schema>,
+    index: usize,
+}
+
+#[cfg(feature = "polars")]
+impl PolarsDataFrameRecordBatchReader {
+    /// Creates a new `PolarsDataFrameRecordBatchReader` from a given Polars DataFrame.
+    pub fn new(df: DataFrame) -> Self {
+        let fields: Vec<arrow_schema::Field> = df
+            .schema()
+            .into_iter()
+            .map(|(name, dtype)| {
+                arrow_schema::Field::new(
+                    name,
+                    arrow_schema::DataType::from(dtype.to_arrow(false)),
+                    true,
+                )
+            })
+            .collect();
+        // Use pl_flavor = false to use LargeBinary and LargeUtf8 Arrow types instead of
+        // BinaryView and Utf8View types because polars-arrow to arrow-rs conversion
+        // is not yet implemented for BinaryView and Utf8View.
+        // See: https://lists.apache.org/thread/w88tpz76ox8h3rxkjl4so6rg3f1rv7wt for the
+        // differences in the types.
+        PolarsDataFrameRecordBatchReader {
+            chunks: df.iter_chunks(false).collect(),
+            arrow_schema: Arc::new(arrow_schema::Schema::new(fields)),
+            index: 0,
+        }
+    }
+}
+
+#[cfg(feature = "polars")]
+impl Iterator for PolarsDataFrameRecordBatchReader {
+    type Item = std::result::Result<arrow_array::RecordBatch, arrow_schema::ArrowError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.chunks.len() {
+            let columns: Vec<arrow_array::ArrayRef> = self.chunks[self.index]
```

Comment:
Instead of storing the chunks as `Vec` and then walking through the `Vec` with an index can we store the chunks as an iterator?  E.g. either...

```
pub struct PolarsDataFrameRecordBatchReader<I: Iterator<Item = ArrowChunk>> {
    chunks: I,
    ...
```

Or just be explicit...

```
pub struct PolarsDataFrameRecordBatchReader {
    chunks: std::vec::IntoIter<ArrowChunk>
```

The advantage here is that chunks are purged from RAM as soon as they are processed.

Reply:
Done. Had to do option 2 - I'm not able to pull the RecordBatchIter struct which implements Iterator<Item = ArrowChunk> due to its lifetime. If we wanted to avoid the allocation into a vector, I think the dataframe would have to outlive the PolarsDataFrameRecordBatchReader struct, which we don't want.